### ISSUE TITLE
value_and_grad(odeint(...)) doesn't work

### ISSUE THREAD

**Issue by samuela** [google/jax#2345]
I'm attempting to get gradients through `odeint`. I have the following code,

<details>
  <summary>Repro code</summary>

  ```python
    import jax
    from jax import random
    from jax import value_and_grad
    import jax.numpy as jp
    from jax.experimental.stax import Dense
    from jax.experimental import ode

    def pendulum_dynamics(mass: float, length: float, gravity: float,
                          friction: float):
      def f(state, action):
        assert state.shape == (2, )
        assert action.shape == (1, )

        theta = state[0]
        theta_dot = state[1]
        u = action[0]

        theta_dotdot = (-gravity / length * jp.sin(theta) - friction * theta_dot +
                        u / (mass * length**2))

        return jp.array([theta_dot, theta_dotdot])

      return f

    def policy_integrate(dynamics, cost, policy):
      def ofunc(y, _t, policy_params):
        x = y[1:]
        u = policy(policy_params, x)
        return jp.concatenate((cost(x, u).reshape((1, )), dynamics(x, u)))

      def eval_policy(policy_params, x0, total_time):
        return ode.odeint(ofunc, jp.concatenate((jp.zeros((1, )), x0)),
                          jp.array([total_time]), policy_params)[0, 0]

      return eval_policy

    def main():
      dynamics = pendulum_dynamics(
          mass=0.1,
          length=1.0,
          gravity=9.8,
          friction=0.0,
      )

      def cost(x, u):
        return (x[0] - jp.pi)**2 + 0.1 * (x[1]**2) + 0.001 * (u[0]**2)

      policy_init, policy = Dense(1)
      _, policy_params = policy_init(random.PRNGKey(0), (2, ))

      run = value_and_grad(policy_integrate(dynamics, cost, policy))
      run(policy_params, jp.zeros((2, )), total_time=30.0)

    if __name__ == "__main__":
      print(jax.__version__)
      main()

  ```
</details>

which produces,
```
0.1.59
/Users/skainswo/.local/share/virtualenvs/research-CJWjSlE2/lib/python3.7/site-packages/jax/lib/xla_bridge.py:122: UserWarning: No GPU/TPU found, falling back to CPU.
  warnings.warn('No GPU/TPU found, falling back to CPU.')
Traceback (most recent call last):
  File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/skainswo/dev/research/research/diffphysics/jaxbug.py", line 56, in <module>
    main()
  File "/Users/skainswo/dev/research/research/diffphysics/jaxbug.py", line 52, in main
    run(policy_params, jp.zeros((2, )), total_time=30.0)
  File "/Users/skainswo/.local/share/virtualenvs/research-CJWjSlE2/lib/python3.7/site-packages/jax/api.py", line 407, in value_and_grad_f
    ans, vjp_py = vjp(f_partial, *dyn_args)
  File "/Users/skainswo/.local/share/virtualenvs/research-CJWjSlE2/lib/python3.7/site-packages/jax/api.py", line 1285, in vjp
    out_primal, out_vjp = ad.vjp(flat_fun, primals_flat)
  File "/Users/skainswo/.local/share/virtualenvs/research-CJWjSlE2/lib/python3.7/site-packages/jax/interpreters/ad.py", line 119, in vjp
    out_primals, pvals, jaxpr, consts = linearize(traceable, *primals)
  File "/Users/skainswo/.local/share/virtualenvs/research-CJWjSlE2/lib/python3.7/site-packages/jax/interpreters/ad.py", line 110, in linearize
    assert all(aval_primal is None for aval_primal in aval_primals)
AssertionError
```

I see that there's
* `my_odeint_grad`
* `my_odeint_jacrev`
* `vjp_odeint`
* `build_odeint`

in `jax.experimental.ode` and it's not clear to me if I should be using one of these instead. If so, I'm not 100% clear on what all the differences are. Am I doing something wrong? What's the correct approach to getting forward/reverse mode gradients out of `odeint`? (In this particular case I'm looking for vjp.)

**Comment by mattjj** [comment]
Thanks for raising this, Sam!

I think it's an issue with the custom VJP rule machinery, which I aim to fix by merging #2026. That PR also includes a rewrite of odeint.

Any chance you could experiment on that branch and see if the issue persists? If it does, I'll fix it there.

**Comment by samuela** [comment]
Hey Matt! I'm working with pipenv right now so checking out a branch may take me a bit of time. I'll give it a shot.

I ended up settling on `vjp_odeint`, but now I'm running into some new issues. In particular, it looks like the VJP doesn't work with only a single `t` specified:

```python
import jax
from jax import random
import jax.numpy as jp
from jax.experimental.stax import Dense
from jax.experimental import ode

def pendulum_dynamics(mass: float, length: float, gravity: float,
                      friction: float):
  def f(state, action):
    assert state.shape == (2, )
    assert action.shape == (1, )

    theta = state[0]
    theta_dot = state[1]
    u = action[0]

    theta_dotdot = (-gravity / length * jp.sin(theta) - friction * theta_dot +
                    u / (mass * length**2))

    return jp.array([theta_dot, theta_dotdot])

  return f

def policy_integrate_and_grad(dynamics, cost, policy):
  def ofunc(y, _t, policy_params):
    x = y[1:]
    u = policy(policy_params, x)
    return jp.concatenate((jp.expand_dims(cost(x, u), axis=0), dynamics(x, u)))

  def value_and_grad(policy_params, x0, total_time):
    y0 = jp.concatenate((jp.zeros((1, )), x0))

    # !!!!!!!!!!!!!!!!!!!!!!!!!
    t = jp.array([total_time])  # doesn't work
    t = jp.array([0.0, total_time])  # works (I hope?)

    primals, vjp = ode.vjp_odeint(ofunc, y0, t, policy_params)
    g = vjp(
        jp.expand_dims(jp.concatenate((jp.ones((1, )), jp.zeros_like(x0))),
                       axis=0))
    return primals, g

  return value_and_grad

def main():
  dynamics = pendulum_dynamics(
      mass=0.1,
      length=1.0,
      gravity=9.8,
      friction=0.0,
  )

  def cost(x, u):
    assert x.shape == (2, )
    assert u.shape == (1, )
    return (x[0] - jp.pi)**2 + 0.1 * (x[1]**2) + 0.001 * (u[0]**2)

  policy_init, policy = Dense(1)
  _, policy_params = policy_init(random.PRNGKey(0), (2, ))

  run = policy_integrate_and_grad(dynamics, cost, policy)
  print(run(policy_params, jp.zeros((2, )), total_time=1.0))

if __name__ == "__main__":
  print(jax.__version__)
  main()
```

I'm getting an error:
```
0.1.59
/Users/skainswo/.local/share/virtualenvs/research-CJWjSlE2/lib/python3.7/site-packages/jax/lib/xla_bridge.py:122: UserWarning: No GPU/TPU found, falling back to CPU.
  warnings.warn('No GPU/TPU found, falling back to CPU.')
Traceback (most recent call last):
  File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/skainswo/dev/research/research/diffphysics/jaxbug.py", line 65, in <module>
    main()
  File "/Users/skainswo/dev/research/research/diffphysics/jaxbug.py", line 61, in main
    print(run(policy_params, jp.zeros((2, )), total_time=1.0))
  File "/Users/skainswo/dev/research/research/diffphysics/jaxbug.py", line 39, in value_and_grad
    axis=0))
  File "/Users/skainswo/.local/share/virtualenvs/research-CJWjSlE2/lib/python3.7/site-packages/jax/experimental/ode.py", line 369, in <lambda>
    vjp_fun = lambda g: vjp_all(g, primals_out, t)
  File "/Users/skainswo/.local/share/virtualenvs/research-CJWjSlE2/lib/python3.7/site-packages/jax/api.py", line 146, in f_jitted
    name=flat_fun.__name__)
  File "/Users/skainswo/.local/share/virtualenvs/research-CJWjSlE2/lib/python3.7/site-packages/jax/core.py", line 642, in call_bind
    outs = primitive.impl(f, *args, **params)
  File "/Users/skainswo/.local/share/virtualenvs/research-CJWjSlE2/lib/python3.7/site-packages/jax/interpreters/xla.py", line 448, in _xla_call_impl
    compiled_fun = _xla_callable(fun, device, backend, name, *map(arg_spec, args))
  File "/Users/skainswo/.local/share/virtualenvs/research-CJWjSlE2/lib/python3.7/site-packages/jax/linear_util.py", line 220, in memoized_fun
    ans = call(fun, *args)
  File "/Users/skainswo/.local/share/virtualenvs/research-CJWjSlE2/lib/python3.7/site-packages/jax/interpreters/xla.py", line 501, in _xla_callable
    extend_name_stack(wrap_name(name, 'jit')), *xla_args)
  File "/Users/skainswo/.local/share/virtualenvs/research-CJWjSlE2/lib/python3.7/site-packages/jax/interpreters/xla.py", line 328, in jaxpr_subcomp
    *in_nodes, backend=backend, **new_params)
  File "/Users/skainswo/.local/share/virtualenvs/research-CJWjSlE2/lib/python3.7/site-packages/jax/lax/lax_control_flow.py", line 247, in _while_loop_translation_rule
    extend_name_stack(name_stack, 'body'), *(y + z))
  File "/Users/skainswo/.local/share/virtualenvs/research-CJWjSlE2/lib/python3.7/site-packages/jax/interpreters/xla.py", line 338, in jaxpr_subcomp
    name_stack, backend=backend, **new_params)
  File "/Users/skainswo/.local/share/virtualenvs/research-CJWjSlE2/lib/python3.7/site-packages/jax/interpreters/xla.py", line 612, in _xla_call_translation_rule
    extend_name_stack(name_stack, wrap_name(name, 'jit')), *args)
  File "/Users/skainswo/.local/share/virtualenvs/research-CJWjSlE2/lib/python3.7/site-packages/jax/interpreters/xla.py", line 343, in jaxpr_subcomp
    c.GetShape(ans)  # force xla to do shape error checking
  File "/Users/skainswo/.local/share/virtualenvs/research-CJWjSlE2/lib/python3.7/site-packages/jaxlib/xla_client.py", line 735, in GetShape
    return self._builder.GetShape(operand)
RuntimeError: Invalid argument: Slice size at index 0 in gather op is out of range, must be within [0, 1), got 1.
```

**Comment by samuela** [comment]
Ok, #2026 and using more than one `t` works!

**Comment by mattjj** [comment]
Merged #2026, so this should be fixed!
